spec/services/waste_exemptions_engine/confirmation_pdf_generator_service_spec.rb

```diff
@@ -4,18 +4,71 @@
 
 module WasteExemptionsEngine
   RSpec.describe ConfirmationPdfGeneratorService do
-    describe ".send_confirmation_email" do
+    describe ".run" do
       it "generates and return a string containing PDF content with the confirmation information" do
-        Timecop.freeze(Time.local(2018, 1, 1, 1, 5, 0))
-        registration = build(:registration, :confirmable)
-        fixtures_file_path = Rails.root.join("..", "fixtures/pdfs/confirmation.pdf")
+        registration = create(:registration, :complete)
 
-        result = ConfirmationPdfGeneratorService.run(registration: registration)
+        result = described_class.run(registration: registration)
 
-        ## Generate new fixtures PDF when a change has been made
-        # File.open(fixtures_file_path, "wb") { |file| file << result }
+        expect(result).to include_pdf_content(registration.reference)
+        expect(result).to include_pdf_content(registration.submitted_at.to_formatted_s(:day_month_year))
+        expect(result).to include_pdf_content("Limited company")
+        expect(result).to include_pdf_content(registration.applicant_phone)
+        expect(result).to include_pdf_content(registration.applicant_email)
+        expect(result).to include_pdf_content(registration.applicant_first_name)
+        expect(result).to include_pdf_content(registration.applicant_last_name)
+        expect(result).to include_pdf_content(registration.operator_name)
+        expect(result).to include_pdf_content(registration.contact_phone)
+        expect(result).to include_pdf_content(registration.contact_email)
+        expect(result).to include_pdf_content(registration.contact_first_name)
+        expect(result).to include_pdf_content(registration.contact_last_name)
+        expect(result).to include_pdf_content(registration.contact_position)
 
-        expect(result).to match_pdf_content(fixtures_file_path)
+        registration.addresses.each do |address|
+          if address.manual?
```

Comment:
It's not a great name 😩 but if the address mode is `auto` then we know we're dealing with the site address in its grid reference and description configuration.

So I think it would be more accurate to say

```ruby
  if address.auto?
    expect(result).to include_pdf_content(address.description)
    expect(result).to include_pdf_content(address.grid_reference)
  else
    expect(result).to include_pdf_content(address.premises)
    expect(result).to include_pdf_content(address.street_address)
    expect(result).to include_pdf_content(address.locality)
    expect(result).to include_pdf_content(address.city)
    expect(result).to include_pdf_content(address.postcode)
  end
```

Reply:
Make sense! What if we create a `grid_reference` method in the `address` class which for now will just check that the address mode is `auto`? Will make it much clearer IMO

Reply:
I get what you mean and 100% agree. I'm just not sure of the name as we have a grid_reference attribute. So this could mean return me that, or has grid reference has been set or not.

So none of these are any better but `uses_grid_reference?`, `located_by_grid_reference?`, `address_is_a_grid_reference?` I think give a bit more indication of the intent/purpose of the method.

I just hope you can do better than my attempts!